File: combination_sum.py
class Solution:
    def combinationSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        res = []

        def dfs(i, curr, total):
            if total == target:
                res.append(curr.copy())
                return

            for j in range(i, len(nums)):
                if total > target:
                    return
                curr.append(nums[j])
                dfs(j, curr, total + nums[j])
                curr.pop()

        dfs(0, [], 0)
        return res

File: combinationSum.js
class Solution {
  /**
   * @param {number[]} nums
   * @param {number} target
   * @returns {number[][]}
   */
  combinationSum(nums, target) {
    nums.sort((a, b) => a - b);
    let res = [];

    const dfs = (i, cur, total) => {
      if (total === target) {
        res.push([...cur]);
        return;
      }

      for (let j = i; j < nums.length; j++) {
        if (total > target) {
          return;
        }
        cur.push(nums[j]);
        dfs(j, cur, total + nums[j]);
        cur.pop();
      }
    };

    dfs(0, [], 0);
    return res;
  }
}
